docs/cplusplus-style-guide.md

```diff
@@ -0,0 +1,372 @@
+  // ...
+
+private:
+  std::size_t size_    = 0;
+  int *       pointer_ = nullptr;
+};
+```
+
+* Mark all constructors and assignment operators explicitly, even those which are implicit, to ensure there is no confusion about the intended behaviour of the class.
+* Mark unneeded default implementations as deleted. C++ infers whether to create implicit constructors based on whether there are user defined ones. This means that when someone declares a constructor, C++ may remove the default constructor.
+
+``` c++ class Transaction
+{
+public:
+  Transaction()                    = default;
+  Transaction(Transaction const &) = default;
+  Transaction(Transaction &&)      = deleted;
+
+  ~Transaction() = default;
+
+  Transaction operator(Transaction const &) = default;
+  Transaction operator(Transaction &&)      = deleted;
+}
+```
+
+* Mark a class `final` if it is not meant to be subclassed. The same counts for final implementations of virtual functions.
+
+``` c++
+class Foo
+{
+public:
+  virtual ~Foo()                        = default;
+  virtual void Greet()                  = 0;
+  virtual void Say(std::string const &) = 0;
+};
+
+class Bar : public Foo
+{
+public:
+  void Greet() override final
+  {
+    // ...
+  }
+
+  void Say(std::string const &what) override {
+      // ...
+  };
+};
+
+class Baz final : public Bar
+{
+public:
+  void Say(std::string const &what) override
+  {
+    // ...
+  }
+};
+```
+
+
+## Error handling
+
+* Always check error codes and return codes.
+* Use exceptions *only* for exceptional circumstances.
+* Handle normal errors gracefully.
+
+* Do not rely on exceptions to avoid taking responsibility.
+* Do not throw exceptions in destructors as this can crash the program.
+
+
+## Memory handling
+
+* Use references where the value cannot be null. Use pointers where it can.
+* Make sure no uninitialised memory is accessed.
+* Always prefer passing by reference or `const` reference where possible.
+* Prefer smart pointers (`std::shared_ptr`, `std::unique_ptr`, etc) over raw pointers.
+* Use `make_shared` and `make_unique` to wrap objects in smart pointers.
+* Prefer the type-safe `nullptr` over the weakly-typed literal `0` and the `NULL` macro.
+* Stick to RAII principle and automatic memory management - destructors are ideally empty.
+* When working with array object types, prefer existing structures that already implement memory management:
+  * From the standard library, such as `std::vector` and `std::string`.
+  * `SharedArray` and `Array` classes from `fetch::memory` namespace.
+  * Classes from `fetch::byte_array` namespace, such as `BasicByteArray`.
+* It is safe to transfer ownership from `std::unique_ptr` to `std::shared_ptr`. However, do **NOT** transfer ownership the other way around, as that will result in memory access exceptions; in this case a non-owning `std::weak_ptr` should be used instead of `std::unique_ptr` (see [reference documentation](https://en.cppreference.com/w/cpp/memory/weak_ptr)).
```

Comment:
I don't think people will try to do this, I would suggest removing lines 315 - 354